cmd/util/ledger/migrations/cadence_values_migration.go

```diff
@@ -0,0 +1,243 @@
+		migrationRuntime.Storage,
+	)
+
+	capabilityIDs := map[interpreter.AddressPath]interpreter.UInt64Value{}
+
+	reporter := newValueMigrationReporter(m.reporter)
+
+	m.log.Info().Msg("Migrating cadence values")
+
+	migration.Migrate(
+		&migrations.AddressSliceIterator{
+			Addresses: []common.Address{
+				address,
+			},
+		},
+		migration.NewValueMigrationsPathMigrator(
+			reporter,
+			&capcons.CapabilityValueMigration{
+				CapabilityIDs: capabilityIDs,
+				Reporter:      reporter,
+			},
+			&capcons.LinkValueMigration{
+				CapabilityIDs:      capabilityIDs,
+				AccountIDGenerator: &AccountIDGenerator{},
+				Reporter:           reporter,
+			},
+			string_normalization.NewStringNormalizingMigration(),
+			account_type.NewAccountTypeMigration(),
+			type_value.NewTypeValueMigration(),
+		),
+	)
+
+	m.log.Info().Msg("Committing changes")
+	migration.Commit()
+
+	// finalize the transaction
+	result, err := migrationRuntime.TransactionState.FinalizeMainTransaction()
+	if err != nil {
+		return nil, fmt.Errorf("failed to finalize main transaction: %w", err)
+	}
+
+	// Merge the changes to the original payloads.
+	return m.mergeRegisterChanges(migrationRuntime, result.WriteSet)
+}
+
+func (m *CadenceValueMigrator) mergeRegisterChanges(
+	mr *migratorRuntime,
+	changes map[flow.RegisterID]flow.RegisterValue,
+) ([]*ledger.Payload, error) {
+
+	originalPayloads := mr.Snapshot.Payloads
+	newPayloads := make([]*ledger.Payload, 0, len(originalPayloads))
+
+	for id, value := range originalPayloads {
+		if len(value.Value()) == 0 {
+			// This is strange, but we don't want to add empty values. Log it.
+			m.log.Warn().Msgf("empty value for key %s", id)
+			continue
+		}
+
+		// If the payload had changed, then use the updated payload.
+		if updatedPayload, contains := changes[id]; contains {
+			key := convert.RegisterIDToLedgerKey(id)
+			newPayloads = append(newPayloads, ledger.NewPayload(key, updatedPayload))
+		} else {
+			newPayloads = append(newPayloads, value)
+		}
+	}
+
+	return newPayloads, nil
+}
+
+// AccountIDGenerator for the link-value migration
+type AccountIDGenerator struct {
+	ids map[common.Address]uint64
+}
+
+func (g *AccountIDGenerator) GenerateAccountID(address common.Address) (uint64, error) {
+	if g.ids == nil {
+		g.ids = make(map[common.Address]uint64)
+	}
+	g.ids[address]++
+	return g.ids[address], nil
+}
+
+// cadenceValueMigrationReporter is the reporter for cadence value migrations
+type cadenceValueMigrationReporter struct {
+	rw reporters.ReportWriter
+}
+
+var _ capcons.LinkMigrationReporter = &cadenceValueMigrationReporter{}
+var _ capcons.CapabilityMigrationReporter = &cadenceValueMigrationReporter{}
+
+func newValueMigrationReporter(rw reporters.ReportWriter) *cadenceValueMigrationReporter {
+	return &cadenceValueMigrationReporter{
+		rw: rw,
+	}
+}
+
+func (t *cadenceValueMigrationReporter) Report(
+	addressPath interpreter.AddressPath,
+	migration string,
+) {
+	t.rw.Write(cadenceValueMigrationReportEntry{
```

Comment:
I don't think I ever tried writing multiple different types to the same report writer. does this work ok?

Reply:
Yes, It really depends on the implementation of the underlying writer.
The `ReportFileWriter` writes to a file as a JSON array. As per JSON spec, an element of an array can be any JSON (object, array, literal, etc.). So different-shaped objects are OK in this case.
I also tested this manually just to verify, and it works fine.